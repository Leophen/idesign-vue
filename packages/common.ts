import _ from 'lodash'
import {
  Component,
  getCurrentInstance,
  Slots,
  VNode,
} from 'vue'

// 通用位置类型
export type placementType =
  | 'top'
  | 'top-left'
  | 'top-right'
  | 'bottom'
  | 'bottom-left'
  | 'bottom-right'
  | 'left'
  | 'left-top'
  | 'left-bottom'
  | 'right'
  | 'right-top'
  | 'right-bottom'

// 通用触发类型
export type triggerType = 'hover' | 'click' | 'context-menu'

// 转换 width/height/left/... 样式值
export const turnValue = (val: string | number | undefined) => {
  if (!val) {
    return
  }
  if (_.isNumber(val as number)) {
    return `${val}px`
  }
  return val
}

// const _stateValue = ref(defaultValue)
// const innerValue = computed(() => value ?? _stateValue.value)

/**
 * 渲染 default slot，获取子组件 VNode。处理多种子组件创建场景
 * @returns {function(childComponentName: string, slots: Slots): VNode[]}
 * @param childComponentName
 * @param slots
 * @example const getChildByName = useChildComponentSlots()
 * @example getChildComponentByName('xxItem')
 */
export const useChildComponentSlots = () => {
  const instance = getCurrentInstance()
  return (childComponentName: string, slots?: Slots): VNode[] => {
    if (!slots) {
      slots = instance?.slots
    }
    const content = slots?.default?.() || []

    return (
      content
        .map((item: VNode) => {
          if (item.children && Array.isArray(item.children))
            return item.children
          return item
        })
        .flat()
        // @ts-ignore
        .filter((item: VNode) =>
          (item.type as Component).name?.endsWith(childComponentName)
        ) as VNode[]
    )
  }
}
