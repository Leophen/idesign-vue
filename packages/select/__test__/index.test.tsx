import { Select,SelectItem } from '../index';
import { render } from '../../common'

describe('Select 组件测试', () => {
  it('create', () => {
    const wrapper = render(
      <Select>
        <SelectItem value="item1">选项一</SelectItem>
        <SelectItem value="item2">选项二</SelectItem>
        <SelectItem value="item3">选项三</SelectItem>
      </Select>
    );
    expect(wrapper.element).toMatchSnapshot();
  });

  it('size', () => {
    const wrapper = render(
      <Select size="small">
        <SelectItem value="item1">选项一</SelectItem>
        <SelectItem value="item2">选项二</SelectItem>
        <SelectItem value="item3">选项三</SelectItem>
      </Select>
    );
    expect(wrapper.find('.i-input').classes()).toContain('i-input--size-small');
  });

  it('disabled', () => {
    const wrapper = render(
      <Select disabled={true}>
        <SelectItem value="item1">选项一</SelectItem>
        <SelectItem value="item2">选项二</SelectItem>
        <SelectItem value="item3">选项三</SelectItem>
      </Select>
    );
    expect(wrapper.find('.i-input').classes()).toContain('i-input-is-disabled');
  });
});



