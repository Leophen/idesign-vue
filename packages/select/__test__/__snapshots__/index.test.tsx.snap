// Vitest Snapshot v1

exports[`Select 组件测试 > create 1`] = `
<div
  class="i-select"
>
  
  <div
    class="i-popup__reference"
  >
    
    
    <div
      class="i-input i-input-is-readonly"
    >
      <!---->
      
      <!---->
      
      <input
        class="i-input__inner"
        placeholder="请选择"
        readonly=""
      />
      <!---->
      <i
        class="i-icon icon-TipCloseFill"
      />
      <!---->
      <i
        class="i-icon icon-ArrowDown i-input-suffix-icon"
      />
      <!---->
      <!---->
    </div>
    <!---->
    
    
  </div>
  <!--teleport start-->
  <!--teleport end-->
  
</div>
`;
